client/comms/wsconn.go

```diff
@@ -0,0 +1,275 @@
+	conn.reqMtx.Lock()
+	conn.req[id] = route
+	conn.reqMtx.Unlock()
+}
+
+// FetchRoute fetches the route of the provided request id.
+func (conn *WsConn) FetchRoute(id uint64) (string, error) {
+	conn.reqMtx.RLock()
+	route := conn.req[id]
+	conn.reqMtx.RUnlock()
+
+	if route == "" {
+		return "", fmt.Errorf("no route found for id #%v", id)
+	}
+
+	conn.reqMtx.Lock()
+	delete(conn.req, id)
+	conn.reqMtx.Unlock()
+
+	return route, nil
+}
+
+// NextID returns the next request id.
+func (conn *WsConn) NextID() uint64 {
+	return atomic.AddUint64(&conn.rID, 1)
+}
+
+// connect attempts to establish a websocket connection.
+func (conn *WsConn) connect() error {
+	url := url.URL{
+		Scheme: "ws",
+		Host:   conn.host,
+		Path:   "ws",
+	}
+
+	var err error
+	if conn.tlsCfg != nil {
+		dialer := &websocket.Dialer{
+			TLSClientConfig: conn.tlsCfg,
+		}
+		conn.ws, _, err = dialer.Dial(url.String(), nil)
+		if err != nil {
+			return err
+		}
+	} else {
+		conn.ws, _, err = websocket.DefaultDialer.Dial(url.String(), nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	conn.setConnected(true)
+	conn.ws.SetPingHandler(func(data string) error {
+		log.Tracef("Received ping")
+
+		now := time.Now()
+		err := conn.ws.SetReadDeadline(
+			now.Add(time.Duration(conn.pingWait) * time.Second))
+		if err != nil {
+			log.Errorf("unable to set read deadline: %v\n", err)
+			return err
+		}
+
+		// Respond with a pong.
+		return conn.ws.WriteControl(websocket.PongMessage, []byte{}, time.Time{})
+	})
+
+	return nil
+}
+
+// keepAlive maintains an active websocket connection by reconnecting when
+// the established connection is broken. This should be run as a goroutine.
+func (conn *WsConn) keepAlive(ctx context.Context) {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			connected := conn.fetchConnected()
+			if !connected {
+				log.Trace("reconnecting client")
+				err := conn.connect()
+				if err != nil {
+					log.Errorf("unable to establish a connection: %v\n", err)
+				}
+			}
+
+		case <-ctx.Done():
+			conn.ws.Close()
+			conn.wg.Done()
+			return
+		}
+	}
+}
+
+// read fetches and parses incoming messages for processing. This should be
+// run as a goroutine.
+func (conn *WsConn) read() {
+	for {
+		// Only allow reads if there is an established connection.
+		connected := conn.fetchConnected()
+		if !connected {
+			time.Sleep(time.Millisecond * 500)
+			continue
+		}
+
+		_, data, err := conn.ws.ReadMessage()
+		if err != nil {
+			if websocket.IsCloseError(err, websocket.CloseAbnormalClosure, websocket.CloseNormalClosure) ||
+				strings.Contains(err.Error(), "use of closed network connection") {
+				conn.setConnected(false)
+				continue
+			}
+
+			if nErr := err.(net.Error); nErr != nil {
+				if nErr.Timeout() {
+					conn.setConnected(false)
+					continue
+				}
+			}
+
+			log.Errorf("unable to read message: %v\n", err)
+			return
+		}
+
+		log.Tracef("read message: %v", spew.Sdump(data))
```

Comment:
We need to be careful about logging client messages, which can be anything and almost any size.  Also, even if the log level does not include trace messages, this will still call `spew.Sdump`.  Also, since `data` is a `[]byte`, I'm not sure what spew would do differently from just printing like `%v` would print a `[]byte`.

Reply:
spew shows the raw bytes as well as the json message, that's why I'm using it.

Reply:
Understood, I just don't think we want that logged, and we don't want to call `Sdump` every time (still gets called regardless of log level).

Reply:
Will strip out the trace messages for now.